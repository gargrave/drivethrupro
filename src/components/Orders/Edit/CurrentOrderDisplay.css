.App .current-order-total-price {
  margin-left: 1.5rem;
}
