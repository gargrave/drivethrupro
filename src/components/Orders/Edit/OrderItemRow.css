.App .order-item-row .order-item-row-item + .order-item-row-item {
  margin-left: 30px;
  margin-top: 5px;
}
