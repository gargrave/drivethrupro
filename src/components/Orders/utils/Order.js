/*
A simple utility class for keeping track of details relating to an order.

If an existing Order object is passed, its data will be copied into this one;
this is useful for editing existing orders while maintaining the data (particularly the ID).

If nothing is passed in, a new Order will be created with default values.
*/

// A simple "primary key" mock for each object (i.e. to mock the functionality of a DB)
let nextID = 1

export default class Order {
  constructor (otherOrder) {
    if (otherOrder) {
      this.id = otherOrder.id
      this.items = otherOrder.items
      this.totalItems = otherOrder.totalItems
      this.totalPrice = otherOrder.totalPrice
    } else {
      this.id = nextID++
      this.items = []
      this.totalItems = 0
      this.totalPrice = 0
    }
  }

  addItem (orderItem) {
    this.items.push(orderItem)
    this.totalPrice += orderItem.totalPrice
    this.totalItems += orderItem.quantity
  }

  removeItem (orderItem) {
    if (this.items.find(i => i.id === orderItem.id)) {
      this.items = this.items.filter(i => i.id !== orderItem.id)
      this.totalPrice -= orderItem.totalPrice
      this.totalItems -= orderItem.quantity
    }
  }
}
