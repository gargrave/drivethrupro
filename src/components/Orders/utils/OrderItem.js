/*
A simple utility class for keeping track of a single line item within an order.
*/
let nextID = 1

export default class OrderItem {
  constructor (name, quantity, size) {
    this.id = nextID++
    this.name = `${name} (${size.name})`
    this.quantity = quantity
    this.pricePerUnit = size.price
    this.totalPrice = this.quantity * this.pricePerUnit
  }
}
