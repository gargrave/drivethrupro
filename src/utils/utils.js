export function displayAsDollars (cents) {
  if (!Number.isInteger(cents)) {
    throw new TypeError('"cents" must be an Integer.')
  }

  const prefix = cents < 0 ? '-' : ''
  const absValue = Math.abs(cents * 0.01)
  return `${prefix}$${absValue.toFixed(2)}`
}
