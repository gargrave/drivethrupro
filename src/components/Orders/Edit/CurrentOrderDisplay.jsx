import React from 'react'
import { array, func, number, shape } from 'prop-types'

import { displayAsDollars } from '../../../utils/utils'
import OrderItemRow from './OrderItemRow'

const CurrentOrderDisplay = props => {
  return (
    <section>
      <h3>Current Order</h3>
      <h4 className="current-order-total-price">
        Total: {displayAsDollars(props.order.totalPrice)}
      </h4>

      <ul>
        {props.order.items.map(item =>
          <OrderItemRow
            key={item.id}
            item={item}
            handleRemoveItemClick={e => props.handleRemoveItemClick(e, item.id)}
          />
        )}
      </ul>
    </section>
  )
}

CurrentOrderDisplay.propTypes = {
  order: shape({
    items: array.isRequired,
    totalPrice: number.isRequired
  }).isRequired,
  handleRemoveItemClick: func.isRequired
}

export default CurrentOrderDisplay
