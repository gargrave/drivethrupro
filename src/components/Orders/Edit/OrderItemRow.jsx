import React from 'react'
import { func, number, shape, string } from 'prop-types'

import { displayAsDollars } from '../../../utils/utils'
import Button from '../../Common/Button'

import './OrderItemRow.css'

const OrderItemRow = props => {
  return (
    <tr className="order-item-row">
      <td className="order-item-row-item order-item-row-name">
        {props.item.name}
      </td>
      <td className="order-item-row-item order-item-row-quantity">
        {props.item.quantity} @ {displayAsDollars(props.item.pricePerUnit)}
      </td>
      <td className="order-item-row-item order-item-row-price">
        {displayAsDollars(props.item.pricePerUnit * props.item.quantity)}
      </td>
      <td className="order-item-row-item">
        <Button
          inline
          outline
          danger
          className="order-item-row-item remove-item-btn"
          onClick={props.handleRemoveItemClick}
        >
          Remove
        </Button>
      </td>
    </tr>
  )
}

OrderItemRow.propTypes = {
  item: shape({
    name: string.isRequired,
    quantity: number.isRequired,
    pricePerUnit: number.isRequired
  }).isRequired,
  handleRemoveItemClick: func.isRequired
}

export default OrderItemRow
