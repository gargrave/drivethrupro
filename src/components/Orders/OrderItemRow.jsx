import React from 'react'
import { number, shape, string } from 'prop-types'

import { displayAsDollars } from '../../utils/utils'

const OrderItemRow = props => {
  return (
    <li>
      {props.item.name} | {props.item.quantity} | {displayAsDollars(props.item.totalPrice)}
    </li>
  )
}

OrderItemRow.propTypes = {
  item: shape({
    name: string.isRequired,
    quantity: number.isRequired,
    totalPrice: number.isRequired
  }).isRequired
}

export default OrderItemRow
