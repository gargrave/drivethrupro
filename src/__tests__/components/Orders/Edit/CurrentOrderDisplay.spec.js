import React from 'react'
import { shallow } from 'enzyme'

import { displayAsDollars } from '../../../../utils/utils'
import CurrentOrderDisplay from '../../../../components/Orders/Edit/CurrentOrderDisplay'
import OrderItemRow from '../../../../components/Orders/Edit/OrderItemRow'

import Order from '../../../../components/Orders/utils/Order'
import OrderItem from '../../../../components/Orders/utils/OrderItem'

const sizes = [{ name: 'Small', price: 49 }, { name: 'Medium', price: 69 }, { name: 'Large', price: 89 }]

describe('CurrentOrderDisplay', () => {
  let order
  let props
  let component

  beforeEach(() => {
    order = new Order()
    order.addItem(new OrderItem('Item A', 1, sizes[0]))
    order.addItem(new OrderItem('Item B', 2, sizes[1]))

    props = {
      order,
      handleRemoveItemClick: jest.fn()
    }

    component = shallow(<CurrentOrderDisplay {...props} />)
  })

  test('renders correctly', () => {
    expect(component).toMatchSnapshot()
  })

  test('display the correct current price', () => {
    const actual = component.find('.current-order-total-price').text()
    const expected = displayAsDollars(order.totalPrice)
    expect(actual.indexOf(expected)).not.toEqual(-1)
  })

  test('display render the correct number of item rows', () => {
    const actual = component.find(OrderItemRow).length
    const expected = order.items.length
    expect(actual).toEqual(expected)
  })
})
