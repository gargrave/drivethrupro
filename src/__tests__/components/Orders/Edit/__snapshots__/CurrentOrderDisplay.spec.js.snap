// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentOrderDisplay renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <section>
    <h4>
        Current Order:
        <span
            className="current-order-total-price"
        >
            <strong>
                $1.87
            </strong>
        </span>
    </h4>
    <table>
        <thead>
            <tr>
                <th>
                    Product
                </th>
                <th>
                    Quantity
                </th>
                <th>
                    Total Price
                </th>
                <th />
            </tr>
        </thead>
        <tbody>
            <OrderItemRow
                handleRemoveItemClick={[Function]}
                item={
                    OrderItem {
                        "id": 1,
                        "name": "Item A (Small)",
                        "pricePerUnit": 49,
                        "quantity": 1,
                        "totalPrice": 49,
                      }
                }
            />
            <OrderItemRow
                handleRemoveItemClick={[Function]}
                item={
                    OrderItem {
                        "id": 2,
                        "name": "Item B (Medium)",
                        "pricePerUnit": 69,
                        "quantity": 2,
                        "totalPrice": 138,
                      }
                }
            />
        </tbody>
    </table>
</section>,
  "nodes": Array [
    <section>
      <h4>
            Current Order:
            <span
                  className="current-order-total-price"
            >
                  <strong>
                        $1.87
                  </strong>
            </span>
      </h4>
      <table>
            <thead>
                  <tr>
                        <th>
                              Product
                        </th>
                        <th>
                              Quantity
                        </th>
                        <th>
                              Total Price
                        </th>
                        <th />
                  </tr>
            </thead>
            <tbody>
                  <OrderItemRow
                        handleRemoveItemClick={[Function]}
                        item={
                              OrderItem {
                                    "id": 1,
                                    "name": "Item A (Small)",
                                    "pricePerUnit": 49,
                                    "quantity": 1,
                                    "totalPrice": 49,
                                  }
                        }
                  />
                  <OrderItemRow
                        handleRemoveItemClick={[Function]}
                        item={
                              OrderItem {
                                    "id": 2,
                                    "name": "Item B (Medium)",
                                    "pricePerUnit": 69,
                                    "quantity": 2,
                                    "totalPrice": 138,
                                  }
                        }
                  />
            </tbody>
      </table>
</section>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <CurrentOrderDisplay
        handleRemoveItemClick={[Function]}
        order={
                Order {
                        "id": 1,
                        "items": Array [
                          OrderItem {
                            "id": 1,
                            "name": "Item A (Small)",
                            "pricePerUnit": 49,
                            "quantity": 1,
                            "totalPrice": 49,
                          },
                          OrderItem {
                            "id": 2,
                            "name": "Item B (Medium)",
                            "pricePerUnit": 69,
                            "quantity": 2,
                            "totalPrice": 138,
                          },
                        ],
                        "totalItems": 3,
                        "totalPrice": 187,
                      }
        }
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "handleRemoveItemClick": [Function],
          "order": Order {
            "id": 1,
            "items": Array [
              OrderItem {
                "id": 1,
                "name": "Item A (Small)",
                "pricePerUnit": 49,
                "quantity": 1,
                "totalPrice": 49,
              },
              OrderItem {
                "id": 2,
                "name": "Item B (Medium)",
                "pricePerUnit": 69,
                "quantity": 2,
                "totalPrice": 138,
              },
            ],
            "totalItems": 3,
            "totalPrice": 187,
          },
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <section>
          <h4>
                    Current Order:
                    <span
                              className="current-order-total-price"
                    >
                              <strong>
                                        $1.87
                              </strong>
                    </span>
          </h4>
          <table>
                    <thead>
                              <tr>
                                        <th>
                                                  Product
                                        </th>
                                        <th>
                                                  Quantity
                                        </th>
                                        <th>
                                                  Total Price
                                        </th>
                                        <th />
                              </tr>
                    </thead>
                    <tbody>
                              <OrderItemRow
                                        handleRemoveItemClick={[Function]}
                                        item={
                                                  OrderItem {
                                                            "id": 1,
                                                            "name": "Item A (Small)",
                                                            "pricePerUnit": 49,
                                                            "quantity": 1,
                                                            "totalPrice": 49,
                                                          }
                                        }
                              />
                              <OrderItemRow
                                        handleRemoveItemClick={[Function]}
                                        item={
                                                  OrderItem {
                                                            "id": 2,
                                                            "name": "Item B (Medium)",
                                                            "pricePerUnit": 69,
                                                            "quantity": 2,
                                                            "totalPrice": 138,
                                                          }
                                        }
                              />
                    </tbody>
          </table>
</section>,
        "_debugID": 2,
        "_renderedOutput": <section>
          <h4>
                    Current Order:
                    <span
                              className="current-order-total-price"
                    >
                              <strong>
                                        $1.87
                              </strong>
                    </span>
          </h4>
          <table>
                    <thead>
                              <tr>
                                        <th>
                                                  Product
                                        </th>
                                        <th>
                                                  Quantity
                                        </th>
                                        <th>
                                                  Total Price
                                        </th>
                                        <th />
                              </tr>
                    </thead>
                    <tbody>
                              <OrderItemRow
                                        handleRemoveItemClick={[Function]}
                                        item={
                                                  OrderItem {
                                                            "id": 1,
                                                            "name": "Item A (Small)",
                                                            "pricePerUnit": 49,
                                                            "quantity": 1,
                                                            "totalPrice": 49,
                                                          }
                                        }
                              />
                              <OrderItemRow
                                        handleRemoveItemClick={[Function]}
                                        item={
                                                  OrderItem {
                                                            "id": 2,
                                                            "name": "Item B (Medium)",
                                                            "pricePerUnit": 69,
                                                            "quantity": 2,
                                                            "totalPrice": 138,
                                                          }
                                        }
                              />
                    </tbody>
          </table>
</section>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <CurrentOrderDisplay
    handleRemoveItemClick={[Function]}
    order={
        Order {
            "id": 1,
            "items": Array [
              OrderItem {
                "id": 1,
                "name": "Item A (Small)",
                "pricePerUnit": 49,
                "quantity": 1,
                "totalPrice": 49,
              },
              OrderItem {
                "id": 2,
                "name": "Item B (Medium)",
                "pricePerUnit": 69,
                "quantity": 2,
                "totalPrice": 138,
              },
            ],
            "totalItems": 3,
            "totalPrice": 187,
          }
    }
/>,
}
`;
